builtin/sparse-checkout.c

```diff
@@ -259,9 +259,21 @@ static void write_cone_to_file(FILE *fp, struct pattern_list *pl)
 static int write_patterns_and_update(struct pattern_list *pl)
 {
 	char *sparse_filename;
+	struct strbuf backup_filename = STRBUF_INIT;
 	FILE *fp;
+	int sparse_existed;
+	int result;
 
 	sparse_filename = get_sparse_checkout_filename();
+	strbuf_addstr(&backup_filename, sparse_filename);
+	strbuf_addstr(&backup_filename, ".backup");
```

Comment:
Maybe this is overkill but should we be using the `.lock` mechanisms in git to write the change to the `.lock` file and delete that on failure or rename on success?

Reply:
Also thinking about this a little and do we need the protection of the `.lock` to prevent 2 `sparse-checkout` commands from running at the same time?

Reply:
You are probably right about the `.lock` file, but that won't solve the situation here: we need to commit the `.lock` file before the update can work on the new data.

However, I do wonder if we can do something more clever: we have a `struct pattern_list`. Can we try running `clear_ce_flags()` based on this list and get an error from that instead? I will investigate.